### PR TITLE
[SPARK-41858][SQL] Fix ORC reader perf regression due to DEFAULT value feature

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR is a partial and logical revert of SPARK-39862, https://github.com/apache/spark/pull/37280, to fix the huge ORC reader perf regression (3x slower).

SPARK-39862 should propose a fix without perf regression.

### Why are the changes needed?

During Apache Spark 3.4.0 preparation, SPARK-41782 identified a perf regression.
- https://github.com/apache/spark/pull/39301#discussion_r1059239575

### Does this PR introduce _any_ user-facing change?

After this PR, the regression is removed. However, the bug of DEFAULT value feature will remain. This should be handled separately.

### How was this patch tested?

Pass the CI.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala:60]: A little magical,  let me do some investigate too

**Review by LuciferYang — APPROVED** [review]
+1, LGTM if test pass

**Comment by dongjoon-hyun** [comment]
Could you review this, @viirya ?

**Comment by dongjoon-hyun** [comment]
Also, cc @dtenedor , @beliefer , @HyukjinKwon , too

**Review by dtenedor — COMMENTED** [review]
Thanks for doing the debugging investigation, and apologies for the regression! I'm happy to help with the fix if/where needed. I left a comment on the PR here, it looks like it may inadvertently break the DEFAULT value functionality for the Orc data source, possibly causing incorrectness for users. Let's figure it out together and come up with a fix.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: Thank you for review, @dtenedor .
- Please see https://issues.apache.org/jira/browse/SPARK-41782 . We have a benchmark to detect this kind of perf regression. You can run it locally in your environment.
- This is a partial revert to the original code which is the existing behavior before your PR like the previous Spark. As I mentioned in the PR description, [SPARK-39862](https://issues.apache.org/jira/browse/SPARK-39862) should propose a fix without perf regression.

New feature is good as long as not breaking the old behavior.

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: > Thank you for review, @dtenedor .
> 
> * Please see https://issues.apache.org/jira/browse/SPARK-41782 . We have a benchmark to detect this kind of perf regression. You can run it locally in your environment.

Thanks @dongjoon-hyun for the benchmark! The Jira simply comprises the title `Regenerate benchmark results`. Is there some instructions for how to run the benchmark?

> * This is a partial revert to the original code which is the existing behavior before your PR like the previous Spark. As I mentioned in the PR description, [SPARK-39862](https://issues.apache.org/jira/browse/SPARK-39862) should propose a fix without perf regression.
> 
> New feature is good as long as not breaking the old behavior.

Agree on this. However, that bug fix was merged into Spark 3.3 on Jul. 28, 2022. Is it possible that users could have built pipelines since then using the new feature that would return incorrect results if we merged this PR?

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: If someone can tell me how to run the benchmark, I can play around with a fix for the perf regression that also does not change the behavior. I suspect that it's due to the extra function call overhead that takes place each time a value is written, but not sure without a profile :)

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: I copied a wrong link. Here is the exact link, @dtenedor . 
https://github.com/apache/spark/pull/39301#discussion_r1059239575

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: Here is the benchmark.
- https://github.com/apache/spark/blob/master/sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: Please run the above command and check the following case specifically, @dtenedor . 

https://github.com/apache/spark/blob/ec594236df4a568d3b844089b471373789a67b2a/sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcReadBenchmark.scala#L400

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: @dongjoon-hyun sounds good, the comment in that file explains how to run it. Let me take a look myself, I will try to fix the performance regression without changing the behavior. I can ping this thread again with the results

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: Thanks, but I filed SPARK-41858 as the blocker of Apache Spark 3.4 because this really blocks Apache Spark 3.4 preparation from my side. If you don't mind, I'd prefer to merge this first and help you cleanly.

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: @dongjoon-hyun that is OK, but we should probably not release Spark 3.4 with a correctness regression either. It is equally important as performance. If we create another Jira blocking the release of 3.4 that covers fixing the correctness bug, it is fine to merge this PR.

At any rate, I am hoping to figure this out today. Then we should be unblocked 🤔

**Review by viirya — APPROVED** [review]
Looks good to me as this reverts previous patch due to performance regression. Seems there are investigation going on.

Either we can wait for it, or, we can revert it first and apply updated patch later.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: Thank you so much for your understanding. I'll file another blocker JIRA for that of course.

**Review by dtenedor — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: LGTM, let's merge this then

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya , @dtenedor , @beliefer , @LuciferYang . Merged to master.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:1947]: You can assign the new correctness Jira for me to fix.

**Comment by dongjoon-hyun** [comment]
Here is the blocker-level correctness issue JIRA, SPARK-4186, for ORC reader part.

**Comment by beliefer** [comment]
LGTM. Although later.

**Pull request closed without merging**
